### PR TITLE
[DUOS-1408][risk=no] Overhaul of Researcher Profile page

### PR DESCRIPTION
**Pull request by MiriamAlter**
Issue: https://broadworkbench.atlassian.net/browse/DUOS-1408
- Updated ResearcherProfile.js to React hook format
- Added new dynamic text box to inform researchers of empty required fields
- Simplified previous buttons and dialogs to one all-inclusive "Save" button
- Modified API call to back end if user saves an incomplete profile
- Corrected optional eRA Commons ID field to be PI's rather than user's
- Fixed a bug that stopped library cards from rendering after saving
- Department is no longer a required field

----
Have you read [Terra's Contributing Guide](https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/pages/ResearcherProfile.js [hunk 1]
```diff
@@ -1,492 +1,351 @@
...
+    havePI: null,
+    piName: '',
+    piEmail: '',
+    piERACommonsID: '',
+    pubmedID: '',
+    scientificURL: '',
+    completed: undefined
+  });
+
+  const [isLoading, setIsLoading] = useState(true);
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js]: Doesn't look like we use `isLoading` any more

File: src/pages/ResearcherProfile.js [hunk 2]
```diff
@@ -1,492 +1,351 @@
...
+    init();
+  }, [props.history]);
 
-  handleChange = (event) => {
-    let field = event.target.name;
-    let value = event.target.value;
+  useEffect(() => {
+    const isValid = (value) => {
+      let isValid = false;
+      if (trim(value.toString()) !== '' && value !== null && value !== undefined) {
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js]: If value is null or undefined, this will throw an error with `value.toString()`. I suggest using Lodash `isNil` and `isEmpty` methods to simplify this.

Comment (rushtong) [reply src/pages/ResearcherProfile.js]: I misspoke here ... `trim` will handle nil and undefined just fine and return an empty string. Maybe wrap that in `isEmpty`?
```suggestion
      if (!isEmpty(trim(value.toString()))) {
```

File: src/pages/ResearcherProfile.js [hunk 3]
```diff
@@ -1,492 +1,351 @@
...
-      prev.profile[field] = value;
-      return prev;
-    }, () => {
-      if (this.state.validateFields) {
-        this.validateUserFields();
+    const isValidNumber = (value) => {
+      let isValid = false;
+      if (value !== 0 && value !== null && value !== undefined) {
+        isValid = true;
       }
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js]: Lodash is your friend here ... https://lodash.com/docs/4.17.15#isNumber

File: src/pages/ResearcherProfile.js [hunk 4]
```diff
@@ -1,492 +1,351 @@
...
-      profileName = true;
-      showValidationMessages = true;
+    if (!isValid(profile.profileName)) {
+      incompletes.push('Name');
     }
 
-    if (!this.isValid(this.state.profile.academicEmail) || this.state.profile.academicEmail.indexOf('@') === -1) {
-      academicEmail = true;
-      showValidationMessages = true;
+    if (!isValid(profile.academicEmail || profile.academicEmail.indexOf('@') === -1)) {
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js:97]: This is not a request for this PR, but for a separate one in a separate ticket.

We should have a utility method to validate email addresses. See https://stackoverflow.com/questions/46155/how-to-validate-an-email-address-in-javascript for a potential solution to this. I looked in [Terra's codebase](https://github.com/DataBiosphere/terra-ui), but didn't see anything there that we could use.

Comment (rushtong) [reply src/pages/ResearcherProfile.js:97]: This is an example of what I would describe as technical debt. There is a correct way to do a particular task, but it is not a direct requirement of the current task, but instead something that we should do to improve the overall health of the codebase.

File: src/pages/ResearcherProfile.js [hunk 5]
```diff
@@ -1,492 +1,351 @@
...
+      incompletes.push('Email Address');
     }
 
-    if (!this.isValidNumber(this.state.institutionId)) {
-      institution = true;
-      showValidationMessages = true;
+    if (!isValidNumber(profile.institutionId)) {
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js]: See https://lodash.com/docs/4.17.15#isNumber

File: src/pages/ResearcherProfile.js [hunk 6]
```diff
@@ -1,492 +1,351 @@
...
-    });
-  };
+    init();
+  }, [props.history]);
 
-  handleChange = (event) => {
-    let field = event.target.name;
-    let value = event.target.value;
+  useEffect(() => {
+    const isValid = (value) => {
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js]: I don't think you need these *valid functions to be inside useEffect. They can exist outside of useEffect since they are all isolated functions that operate on their arguments and don't modify state in any way. The things you DO want in useEffect are the imcompletes and the checks in each of the incomplete fields.

File: src/pages/ResearcherProfile.js [hunk 7]
```diff
@@ -1,492 +1,351 @@
...
-      return prev;
-    });
-  };
+    newFields[field] = value;
 
-  saveProfile = (event) => {
-    event.preventDefault();
-    this.setState({ showDialogSave: true });
-  };
+    if (profile.isThePI === true || profile.isThePI === 'true') {
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js:221]: This is a great opportunity for something that checks for [truthy-ness](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) but would be smart about the string value `"false"`. Not a request for this PR, but we should have a library method that does something like this in a single call.

File: src/pages/ResearcherProfile.js [hunk 8]
```diff
@@ -1,492 +1,351 @@
...
+      newFields.havePI = '';
+    }
+    if (profile.havePI === false || profile.havePI === 'false' || profile.havePI === '') {
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js:224]: Same comment here, but for [falsy-ness](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) instead.

File: src/pages/ResearcherProfile.js [hunk 9]
```diff
@@ -1,492 +1,351 @@
...
-    this.setState({ showDialogSubmit: false });
-    this.props.history.push({ pathname: 'dataset_catalog' });
+  const eraValidate = async () => {
+    const user = await User.getMe();
+    const userProps = getPropertyValuesFromUser(user);
+    const expirationCount = isNil(userProps.eraExpiration) ? 0 : AuthenticateNIH.expirationCount(userProps.eraExpiration);
+
+    return (!isNil(userProps.eraCommonsId) && userProps.eraAuthorized === 'true' && expirationCount >= 0);
+
+    // Temporary fix until eRACommons.js is updated to share re-render info.
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js]: Let's move the comment to the top of the method with a TODO. This is another example of technical-debt which is acceptable to incur, but should be dealt with in a timely manner. Please feel free to create a ticket for this and start on it if you wish.

File: src/pages/ResearcherProfile.js [hunk 10]
```diff
@@ -1,492 +1,351 @@
...
     }
   };
 
-  updateResearcher = async (profile) => {
-    const profileClone = this.cloneProfile(profile);
-    await Researcher.updateProperties(Storage.getCurrentUser().dacUserId, true, profileClone);
-    await this.saveUser();
-    this.setState({ showDialogSubmit: false });
-    this.props.history.push({ pathname: 'dataset_catalog' });
+  const createResearcher = async (profile) => {
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js]: Let's update this method name to reflect what it's really doing. Naming is hard!  I think it should be:
```suggestion
  const createUserProperties = async (profile) => {
```
Additionally, and separately, `Researcher.createProperties(profile);` should be renamed/moved to `User.createProperties(profile);`

File: src/pages/ResearcherProfile.js [hunk 11]
```diff
@@ -1,492 +1,351 @@
...
+    await Researcher.createProperties(profile);
+    await updateUser();
+    props.history.push({ pathname: 'dataset_catalog' });
   };
 
-  saveUser = async () => {
+  const updateResearcher = async (profile) => {
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js]: Similar naming comment applies here.

File: src/pages/ResearcherProfile.js (hunk 10 above)
Comment (rushtong) [reply src/pages/ResearcherProfile.js]: To be clear, 
> Additionally, and separately, Researcher.createProperties(profile); should be renamed/moved to User.createProperties(profile);

is meant for a separate PR, not this one.

**Review by rushtong — COMMENTED** [review]
This is my first round of review ... I have not tested functionality yet. Please see comments inline. This is a lot of great work, thank you!

**Comment by rushtong** [comment]
It looks like you'll need to merge the latest version of the develop branch into this one.

File: src/pages/ResearcherProfile.js (hunk 1 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js]: I was wondering about that. Fixed

File: src/pages/ResearcherProfile.js (hunk 2 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js]: Fixed and simplified

File: src/pages/ResearcherProfile.js (hunk 6 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js]: The linter recommended I move those functions into useEffect. They don't get used outside of the incomplete checks so I didn't think it would hurt.

File: src/pages/ResearcherProfile.js (hunk 9 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js]: Fixed? Edited, at least

File: src/pages/ResearcherProfile.js (hunk 4 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js:97]: Noted, will post to Jira after next push

File: src/pages/ResearcherProfile.js (hunk 7 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js:221]: Noted, will post to Jira after next push

File: src/pages/ResearcherProfile.js (hunk 8 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js:224]: Noted, will post to Jira after next push

File: src/pages/ResearcherProfile.js (hunk 10 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js]: Fixed, will post issue to Jira after next push

File: src/pages/ResearcherProfile.js (hunk 11 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js]: Fixed

File: src/pages/ResearcherProfile.js (hunk 1 above)
Comment (rushtong) [reply src/pages/ResearcherProfile.js]: Still seeing this line in my latest update.

File: src/pages/ResearcherProfile.js (hunk 6 above)
Comment (rushtong) [reply src/pages/ResearcherProfile.js]: They aren't used outside of the useEffect, that's true, and it's correct that it has no effect on functionality either inside or outside. The linter warning is very strange ... these methods have no dependency on any state values - they're purely functional and could be in an external library if we had more use for them. For example,
```
    const isValidNumber = (value) => {
      let isValid = false;
      if (value !== 0 && value !== null && value !== undefined) {
        isValid = true;
      }
      return isValid;
    };
``` 
Actually can be a call to the lodash function just as easily, so I'm really confused about why the linter is complaining. The point of my comment is that as more things are moved inside of useEffect, it becomes harder to reason about. The things that really need to be there are the state-affecting functions that useEffect has a dependency on, i.e. things that touch `profile` and `incompletes`

File: src/pages/ResearcherProfile.js (hunk 2 above)
Comment (rushtong) [reply src/pages/ResearcherProfile.js]: I'm still seeing the unfixed version, are your updates pushed to origin?

**Review by rushtong — DISMISSED** [review]
Functionally, this is working perfectly! I found a separate bug related to this with incomplete profiles. If a user deletes a required field, and saves, the form works correctly. If the user then navigates to a DAR, the dar form does not flag that the user is missing required fields. Will file a ticket for this.

Once the updates are merged in, please re-request a 👍🏽  and I'll approve it.

**Comment by rushtong** [comment]
@MiriamAlter This branch needs have the latest changes from `develop` merged into this branch, or be rebased onto `develop`. I find the first option easier and more straightforward if there are a lot of commits in your branch.

File: src/pages/ResearcherProfile.js (hunk 6 above)
Comment (JVThomas) [reply src/pages/ResearcherProfile.js]: I'm guessing `isValidState` has to be included because it references the state variable `profile` whose value can change, which is why the linter is stating that it needs to be included. However I don't see linter warnings for the other isValid methods, so those can safely be declared outside the useEffect.

File: src/pages/ResearcherProfile.js (hunk 5 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js]: Fixed

File: src/pages/ResearcherProfile.js (hunk 3 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js]: Fixed

File: src/pages/ResearcherProfile.js (hunk 6 above)
Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js]: Figured out and fixed

File: src/pages/ResearcherProfile.js [hunk 12]
```diff
@@ -525,531 +356,448 @@ export const ResearcherProfile = hh(class ResearcherProfile extends Component {
...
-                      }, countryNames ),
-                      span({
-                        className: 'cancel-color required-field-error-span',
-                        isRendered: this.state.invalidFields.country && showValidationMessages
-                      }, ['Country is required'])
-                    ])
+              div({ className: 'col-xs-12 no-padding' }, [
+                div({ className: 'row fsi-row-lg-level fsi-row-md-level no-margin' }, [
+                  div({ className: 'col-xs-6' }, [
+                    label({ id: 'lbl_profileDepartment', className: 'control-label' }, ['Department']),
```
Comment (rushtong) [inline src/pages/ResearcherProfile.js:518]: It looks like this went from required to non-required with the last update. Can we revert these changes as part of this PR?

Comment (MiriamAlter) [reply src/pages/ResearcherProfile.js:518]: This was part of the discussion in today's standup. Department was changed to non-required on other parts of the site, and they asked me to change it on this page as well.

Comment (rushtong) [reply src/pages/ResearcherProfile.js:518]: Gotcha, thanks, wasn't aware of that 👍🏽

**Review by rushtong — APPROVED** [review]
I'm happy with this, it's a huge overhaul and makes this page a lot easier to work on 👍🏽
